### ISSUE TITLE
Restartable sections support

### ISSUE THREAD

**Issue by andrewmcgr** [facebookincubator/BOLT#90]
Hi,

We're interested in using Bolt in conjunction with restartable sections (see [this article](https://www.efficios.com/blog/2019/02/08/linux-restartable-sequences/#:~:text=Support%20for%20restartable%20sequences%20was,locks%20or%20expensive%20atomic%20instructions.) for details of what restartable sections are).

We're prepared to do code work to support this. My question is, how to do so?

The key property is that a restartable sequence is a range of instructions that must always be handled as a unit. One approach might be to mark functions containing a restartable sequence as non-simple, but that of course means they can't be optimised at all (and they're likely to be hot). Another would be to extend the concept of a basic block such that a block cannot end before the end of the sequence.

What do the Bolt developers think of these? Is there another option? Is anyone already working on this?

Reading the __rseq_cs section seems straightforward, and it looks like the relocations should just work (there are relocations emitted for all the pointers in __rseq_cs).

**Comment by maksfb** [comment]
Great. We don't have anyone working on rseq support actively, but we wanted to add the support because tcmalloc uses rseq's.

In general, since the user can write arbitrary assembly code, it's not easy to support rseq. However, if it's used via library, or in some standardized way, then we can do it. What is your use case?

In terms of implementation, you are right that we need to preserve the original sequence of basic blocks involved in rseq while keeping the ability to reorder basic blocks in the containing function. We don't have an existing  mechanism for it. We will have to modify the reordering algorithm to cluster together basic blocks involved in rseq. This should be pretty straightforward.

**Comment by andrewmcgr** [comment]
The user can indeed write arbitrary assembler, and in a couple of cases we do. However, glibc has a standardised ELF section containing the sequence block pointers, which is done because you can only register for rseq once.

So, while we will be using several different libraries that use rseq, including one of our own, the coexistence requirement is that they all register themselves in __rseq_cs. Which means it would be fine for all our use cases if that were also a requirement for BOLT not to break things.

The ELF section contains an array of these:
```
struct rseq_cs {
    uint version;
    uint flags;
    uint64 start_ip;
    uint64 post_commit_offset;
    uint64 abort_ip;
} __attribute__((aligned(4 * sizeof(uint64))));
```

The critical section starts at `start_ip` and `post_commit_offset` points to the instruction after the end of the section. They're emitted with relocations, so moving sequences around should be possible provided they don't get broken up.

**Comment by andrewmcgr** [comment]
[This comment](https://elixir.bootlin.com/linux/latest/source/kernel/rseq.c#L24) explains the actual constraints, and it looks like the required section runs from a few instructions prior to `start_ip`, which would be inconvenient. I'll investigate some more.

**Comment by maksfb** [comment]
Yes, I  find it confusing that in the comment [1] is not after [start_ip]. It's also different from the example in the article.

**Comment by andrewmcgr** [comment]
Point [1] is where the sequence is 'armed', and it's outside the critical section because it is fine to preempt that store; the kernel need take no action if the PC is still at the store. But if the very next instruction has executed and the kernel preempts, it now has to go do the abort branch.

In any case, we should be able to ensure that we do not break a basic block between or adjacent to the critical section and be safe.

**Comment by andrewmcgr** [comment]
A question: what happens if we have an inline instance of a non-simple function? Is the containing function now non-simple?

**Comment by maksfb** [comment]
BOLT will not inline non-simple functions. If the function was inlined by a compiler, then it depends on the reason it is seen as non-simple by BOLT.
